src/AggregateFunctions/SingleValueData.h

```diff
@@ -0,0 +1,394 @@
+
+private:
+    char small_data[MAX_SMALL_STRING_SIZE]; /// Including the terminating zero.
+
+    char * getDataMutable();
+    const char * getData() const;
+    StringRef getStringRef() const;
+    void allocateLargeDataIfNeeded(UInt32 size_to_reserve, Arena * arena);
+    void changeImpl(StringRef value, Arena * arena);
+
+public:
+    bool has() const override { return size != 0; }
+    void insertResultInto(IColumn & to) const override;
+    void write(WriteBuffer & buf, const ISerialization & /*serialization*/) const override;
+    void read(ReadBuffer & buf, const ISerialization & /*serialization*/, Arena * arena) override;
+
+    bool isEqualTo(const IColumn & column, size_t row_num) const override;
+    bool isEqualTo(const SingleValueDataBase &) const override;
+    void set(const IColumn & column, size_t row_num, Arena * arena) override;
+    void set(const SingleValueDataBase &, Arena * arena) override;
+
+    bool setIfSmaller(const IColumn & column, size_t row_num, Arena * arena) override;
+    bool setIfSmaller(const SingleValueDataBase &, Arena * arena) override;
+
+    bool setIfGreater(const IColumn & column, size_t row_num, Arena * arena) override;
+    bool setIfGreater(const SingleValueDataBase &, Arena * arena) override;
+
+    static bool allocatesMemoryInArena() { return true; }
+};
+
+static_assert(sizeof(SingleValueDataString) == SingleValueDataBase::MAX_STORAGE_SIZE, "Incorrect size of SingleValueDataString struct");
+
+
+/// For any other value types.
+struct SingleValueDataGeneric final : public SingleValueDataBase
+{
+    static constexpr bool is_compilable = false;
+
+private:
+    using Self = SingleValueDataGeneric;
+    Field value;
+
+public:
+    bool has() const override { return !value.isNull(); }
+    void insertResultInto(IColumn & to) const override;
+    void write(WriteBuffer & buf, const ISerialization & serialization) const override;
+    void read(ReadBuffer & buf, const ISerialization & serialization, Arena *) override;
+
+    bool isEqualTo(const IColumn & column, size_t row_num) const override;
+    bool isEqualTo(const SingleValueDataBase & other) const override;
+    void set(const IColumn & column, size_t row_num, Arena *) override;
+    void set(const SingleValueDataBase & other, Arena *) override;
+
+    bool setIfSmaller(const IColumn & column, size_t row_num, Arena * arena) override;
+    bool setIfSmaller(const SingleValueDataBase & other, Arena *) override;
+    bool setIfGreater(const IColumn & column, size_t row_num, Arena * arena) override;
+    bool setIfGreater(const SingleValueDataBase & other, Arena *) override;
+
+    static bool allocatesMemoryInArena() { return false; }
+};
+
+static_assert(sizeof(SingleValueDataGeneric) <= SingleValueDataBase::MAX_STORAGE_SIZE, "Incorrect size of SingleValueDataGeneric struct");
+
+/// min, max, any, anyLast, anyHeavy, etc...
+template <template <typename, bool...> class AggregateFunctionTemplate, bool unary, bool... isMin>
```

Comment:
here `isMin` should have a better general names, as name 'isMin' is just a specific use case for min/max/argMin/argMax, but the SingleValueData is a general one.